clusterloader2/testing/l4ilb/config-ilb-recovery.yaml

```diff
@@ -0,0 +1,134 @@
+# Exec command and when the ILB services are running and reachable.
+# ServiceCreationLatency measurement is used only for determining the moment when all of the
+# ILBs are reachable. Since it reports time based on the initial creation of the service, and
+# not from the start of the action, the measured time will not be shown.
+
+# Constants
+{{$LB_REPLICAS_PER_NS := DefaultParam .CL2_LB_REPLICAS_PER_NS 5}}
+{{$LB_BACKEND_SIZE := DefaultParam .CL2_LB_BACKEND_SIZE 10}}
+{{$EXTERNAL_TRAFFIC_POLICY := DefaultParam .CL2_EXTERNAL_TRAFFIC_POLICY "Cluster"}}
+{{$ilbWaitTimeout := DefaultParam .CL2_ILB_WAIT_TIMEOUT "10m"}}
+{{$ilbQPS := DefaultParam .CL2_ILB_TEST_QPS 20}}
+{{$ILB_RECOVERY_LABEL := "ilb-recovery"}}
+{{$namespaces := 1}}
+
+# Command to be executed
+{{$EXEC_COMMAND := DefaultParam .CL2_EXEC_COMMAND nil}}
+{{$EXEC_TIMEOUT := DefaultParam .CL2_EXEC_TIMEOUT "60m"}}
+{{$EXEC_ADD_ARGUMENT_ONE := DefaultParam .CL2_EXEC_ADD_ARGUMENT_ONE ""}}
+{{$EXEC_ADD_ARGUMENT_TWO := DefaultParam .CL2_EXEC_ADD_ARGUMENT_TWO ""}}
+
+name: l4ilb-recovery
+namespace:
+  number: {{$namespaces}}
+tuningSets:
+- name: ILBConstantQPS
+  qpsLoad:
+    qps: {{$ilbQPS}}
+steps:
+- name: Start measurement for running pods
+  measurements:
+  - Identifier: WaitForRunningDeployments
+    Method: WaitForControlledPodsRunning
+    Params:
+      action: start
+      apiVersion: apps/v1
+      kind: Deployment
+      labelSelector: group = ilb-load
+      operationTimeout: 15m
+- name: Creating ILBs
+  phases:
+  - namespaceRange:
+      min: 1
+      max: {{$namespaces}}
+    replicasPerNamespace: {{$LB_REPLICAS_PER_NS}}
+    tuningSet: ILBConstantQPS
+    objectBundle:
+    - basename: ilb-service
+      objectTemplatePath: service.yaml
+      templateFillMap:
+        DeploymentBaseName: ilb-dep
+        ExternalTrafficPolicy: {{$EXTERNAL_TRAFFIC_POLICY}}
+        ILBSizeLabel: {{$ILB_RECOVERY_LABEL}}
+    - basename: ilb-dep
+      objectTemplatePath: dep.yaml
+      templateFillMap:
+        NumReplicas: {{$LB_BACKEND_SIZE}}
+- name: Waiting for objects creation to be completed
+  measurements:
+  - Identifier: WaitForRunningDeployments
+    Method: WaitForControlledPodsRunning
+    Params:
+      action: gather
+{{if $EXEC_COMMAND}}
+- name: Starting measurements
+  measurements:
+  - Identifier: ServiceCreationLatency
+    Method: ServiceCreationLatency
+    Params:
+      action: start
+      waitTimeout: {{$ilbWaitTimeout}}
+      labelSelector: size = {{$ILB_RECOVERY_LABEL}}
+- name: Starting ILB recovery timer
+  measurements:
+  - Identifier: ILBRecoveryTimer
+    Method: Timer
+    Params:
+      action: start
+      label: {{$ILB_RECOVERY_LABEL}}
+- name: Execute command
+  measurements:
+  - Identifier: ExecCommand
+    Method: Exec
+    Params:
+      timeout: {{$EXEC_TIMEOUT}}
+      command:
+      {{range $EXEC_COMMAND}}
+      - {{.}}
+      {{end}}
+      {{if $EXEC_ADD_ARGUMENT_ONE}}
```

Comment:
After 1:1 discussion: CL2 doesn't currently have support for this

Reply:
- EXEC_COMMAND is included from an override file that contains some fixed values and some that are read from the environment variables.
In my opinion this is the simplest and clearest way to parametrize values that are used in the EXEC command, such as cluster-version to which the node pool should upgrade/downgrade to.
It requires no additional work, but allows us to add specific arguments to some standard commands, so that we can run various tests exactly as required.

- After discussion, we came to a conclusion that making additional arguments into a list is not currently supported. We can do it in the future when it is supported.

Reply:
Thank you.
This allows us to use just one of the EXEC_ADD_ARGUMENTs to add the cluster version, instead of two.

Anyway, I think it’s good to leave both of the arguments until we can make it into a list, since it will let us be more flexible with adding custom values through CL2 parameters. Maybe not for this specific case, but potentially with the same or a different command.
What do you think?